rfc/rfc-61/rfc-61.md

```diff
@@ -0,0 +1,98 @@
+
+## Abstract
+As you know, Hudi already supports basic OCC with abundant lock providers.
+But for multi streaming ingestion writers, the OCC does not work well because the conflicts happen in very high frequency.
+Expand it a little bit, with hashing index, all the writers have deterministic hashing algorithm for distributing the records by primary keys,
+all the keys are evenly distributed in all the data buckets, for a single data flushing in one writer, almost all the data buckets are appended with new inputs,
+so the conflict would very possibility happen for mul-writer because almost all the data buckets are being written by multiple writers at the same time;
+For bloom filter index, things are different, but remember that we have a small file load rebalance strategy to writer into the **small** bucket in higher priority,
+that means, multiple writers prune to write into the same **small** buckets at the same time, that's how conflicts happen.
+
+In general, for multiple streaming writers ingestion, explicit lock is not very capable of putting into production, in this RFC, we propse a lockless solution for streaming ingestion.
+
+## Background
+
+Streaming jobs are naturally suitable for data ingestion, it has no complexity of pipeline orchestration and has a smother write workload.
+Most of the raw data set we are handling today are generating all the time in streaming way.
+
+Based on that, many requests for multiple writers' ingestion are derived. With multi-writer ingestion, several streaming events with the same schema can be drained into one Hudi table,
+the Hudi table kind of becomes a UNION table view for all the input data set. This is a very common use case because in reality, the data sets are usually scattered all over the data sources.
+
+Another very useful use case we wanna unlock is the real-time data set join. One of the biggest pain point in streaming computation is the dataset join,
+the engine like Flink has basic supports for all kind of SQL JOINs, but it stores the input records within its inner state-backend which is a huge cost for pure data join with no additional computations.
+In [HUDI-3304](https://issues.apache.org/jira/browse/HUDI-3304), we introduced a `PartialUpdateAvroPayload`, in combination with the lockless multi-writer,
+we can implement N-ways data sources join in real-time! Hudi would take care of the payload join during compaction service procedure.
+
+## Design
+
+### The Precondition
+
+#### MOR Table Type Is Required
+
+The table type must be `MERGE_ON_READ`, so that we can defer the conflict resolution to the compaction phase. The compaction service would resolve the conflicts of the same keys by respecting the event time sequence of the events.
+
+#### Deterministic Bucketing Strategy
+
+Determistic bucketing strategy is required, because the same records keys from different writers are desired to be distributed into the same bucket, not only for UPSERTs, but also for all the new INSERTs.
```

Comment:
If we are using a MOR table but not a bucket index layout, can we support lockless multi writer for INSERT? Will there be any problem?

Reply:
It will cause multiple writers to write to the same file and cause conflicts.

Reply:
> If we are using a MOR table but not a bucket index layout, can we support lockless multi writer for INSERT? Will there be any problem?
It will cause multiple writers to write to the same file and cause conflicts.

Reply:
Can you explain why the same file is written? In my understanding, different writers will write to different log files, and they will be merged into the same `File Group` during the Compaction phase. Because it's `INSERT` operation, it doesn't need to resolve the same primary key causing conflicts.

Maybe my understanding is wrong, I'd be grateful if you could help me point out.

Reply:
> Can you explain why the same file is written? In my understanding, different writers will write to different log files, and they will be merged into the same `File Group` during the Compaction phase. Because it's `INSERT` operation, it doesn't need to resolve the same primary key causing conflicts.
> 
> Maybe my understanding is wrong, I'd be grateful if you could help me point out.

Sorry, I made a mistake. The new design dose not write the same file. The reason why only bucket index layout can be supported is because, When writing new keys, different writers may assign different filegroups.

Reply:
Yeah, you guys got it, just to ensure that all the changes for one record key happens in one FileGroup.

Reply:
the insert problem can be handled with a more expensive conflict resolution strategy, that incrementally checks for PK violations. let me think about this more and sketch a solution if I can.

Reply:
@danny0405 First of all, this is an exciting change to the concurrency model. Would the `BUCKET` index type also be required for a single ingestion writer in `INSERT` mode and multiple data modification writers (i.e. no new records) in `UPSERT` mode? From my understanding, the `BUCKET` index type would only be required for multiple ingestion writers, specifically in the case of them writing the same _new_ record key. But if the records already exist, the writers modifying them will add deltas to the same file groups no matter the index type.
We've been running into some write conflicts between our ingestion and GDPR deletion; even though our table is partitioned on time, we have some late arrivals that are causing the conflicts. When we created the table, the `BUCKET` index wasn't even available. Given that the index type cannot be changed, it'd be annoying to have to rewrite our entire table (>1PB size). I was hoping we could use the non-blocking concurrency control with the `SIMPLE` index type under the assumption that we have only one ingestion writer that adds new records, and all other writers only modify the existing records. Curious your thoughts on this, and whether it's even something that you'd want to allow in the config since it might be unsafe for most use cases.

Reply:
> I was hoping we could use the non-blocking concurrency control with the SIMPLE index type under the assumption that we have only one ingestion writer that adds new records, and all other writers only modify the existing records.

We can definitely extend the conflict resolution strategy if we have some assumptions like this, so a desired state is any concurrent modifications to the same file group can work smothly with semantics of NB-CC, no matter what the index type is.

In general, thanks for the feedback @psendyk , I think it's a very valuable response with real use cases.